File: 43 Hammurabi/java/src/Hamurabi.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * Game of Hamurabi
 * <p>
 * Based on the Basic game of Hamurabi here
 * https://github.com/coding-horror/basic-computer-games/blob/main/43%20Hammurabi/hammurabi.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's Basic game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 */
public class Hamurabi {

    public static final int INITIAL_POPULATION = 95;
    public static final int INITIAL_BUSHELS = 2800;
    public static final int INITIAL_HARVEST = 3000;
    public static final int INITIAL_LAND_TRADING_AT = 3;
    public static final int INITIAL_CAME_TO_CITY = 5;
    public static final int MAX_GAME_YEARS = 10;
    public static final double MAX_STARVATION_IN_A_YEAR = .45d;

    private int year;
    private int population;
    private int acres;
    private int bushels;
    private int harvest;
    private int landTradingAt;
    private int cameToCity;
    private int starvedInAYear;
    private int starvedOverall;
    private boolean chanceOfPlague;
    private int ratsAte;
    private double peopleFed;
    private double percentageStarved;
    private int bushelsToFeedPeople;

    // Used for keyboard input
    private final Scanner kbScanner;

    private enum GAME_STATE {
        STARTUP,
        INIT,
        YEAR_CYCLE,
        BUY_ACRES,
        SELL_ACRES,
        FEED_PEOPLE,
        PLANT_SEED,
        CALCULATE_HARVEST,
        CALCULATE_BABIES,
        RESULTS,
        FINISH_GAME,
        GAME_OVER
    }

    // Current game state
    private GAME_STATE gameState;

    public Hamurabi() {
        kbScanner = new Scanner(System.in);
        gameState = GAME_STATE.STARTUP;
    }

    /**
     * Main game loop
     */
    public void play() {

        do {
            switch (gameState) {

                case STARTUP:
                    intro();
                    gameState = GAME_STATE.INIT;
                    break;

                case INIT:

                    // These are hard coded startup figures from the basic program
                    year = 0;
                    population = INITIAL_POPULATION;
                    bushels = INITIAL_BUSHELS;
                    harvest = INITIAL_HARVEST;
                    landTradingAt = INITIAL_LAND_TRADING_AT;
                    acres = INITIAL_HARVEST / INITIAL_LAND_TRADING_AT;
                    cameToCity = INITIAL_CAME_TO_CITY;
                    starvedInAYear = 0;
                    starvedOverall = 0;
                    chanceOfPlague = false;
                    ratsAte = INITIAL_HARVEST - INITIAL_BUSHELS;
                    peopleFed = 0;
                    percentageStarved = 0;
                    bushelsToFeedPeople = 0;

                    gameState = GAME_STATE.YEAR_CYCLE;
                    break;

                case YEAR_CYCLE:
                    System.out.println();
                    year += 1;
                    // End of game?
                    if (year > MAX_GAME_YEARS) {
                        gameState = GAME_STATE.RESULTS;
                        break;

                    }
                    System.out.println("HAMURABI:  I BEG TO REPORT TO YOU,");
                    System.out.println("IN YEAR " + year + "," + starvedInAYear + " PEOPLE STARVED," + cameToCity + " CAME TO THE CITY,");
                    population += cameToCity;
                    if (chanceOfPlague) {
                        population /= 2;
                        System.out.println("A HORRIBLE PLAGUE STRUCK!  HALF THE PEOPLE DIED.");
                    }
                    System.out.println("POPULATION IS NOW " + population);
                    System.out.println("THE CITY NOW OWNS " + acres + " ACRES.");
                    System.out.println("YOU HARVESTED " + landTradingAt + " BUSHELS PER ACRE.");
                    System.out.println("THE RATS ATE " + ratsAte + " BUSHELS.");
                    System.out.println("YOU NOW HAVE " + bushels + " BUSHELS IN STORE.");
                    System.out.println();

                    landTradingAt = (int) (Math.random() * 10) + 17;  // Original formula unchanged
                    System.out.println("LAND IS TRADING AT " + landTradingAt + " BUSHELS PER ACRE.");

                    gameState = GAME_STATE.BUY_ACRES;
                    break;

                case BUY_ACRES:
                    int acresToBuy = displayTextAndGetNumber("HOW MANY ACRES DO YOU WISH TO BUY? ");
                    if (acresToBuy < 0) {
                        gameState = GAME_STATE.FINISH_GAME;
                    }

                    if (acresToBuy > 0) {
                        if ((landTradingAt * acresToBuy) > bushels) {
                            notEnoughBushelsMessage();
                        } else {
                            acres += acresToBuy;
                            bushels -= (landTradingAt * acresToBuy);
                            peopleFed = 0;
                            gameState = GAME_STATE.FEED_PEOPLE;
                        }
                    } else {
                        // 0 entered as buy so try to sell
                        gameState = GAME_STATE.SELL_ACRES;
                    }
                    break;

                case SELL_ACRES:
                    int acresToSell = displayTextAndGetNumber("HOW MANY ACRES DO YOU WISH TO SELL? ");
                    if (acresToSell < 0) {
                        gameState = GAME_STATE.FINISH_GAME;
                    }
                    if (acresToSell < acres) {
                        acres -= acresToSell;
                        bushels += (landTradingAt * acresToSell);
                        gameState = GAME_STATE.FEED_PEOPLE;
                    } else {
                        notEnoughLandMessage();
                    }
                    break;

                case FEED_PEOPLE:

                    bushelsToFeedPeople = displayTextAndGetNumber("HOW MANY BUSHELS DO YOU WISH TO FEED YOUR PEOPLE ? ");
                    if (bushelsToFeedPeople < 0) {
                        gameState = GAME_STATE.FINISH_GAME;
                    }

                    if (bushelsToFeedPeople <= bushels) {
                        bushels -= bushelsToFeedPeople;
                        peopleFed = 1;
                        gameState = GAME_STATE.PLANT_SEED;
                    } else {
                        notEnoughBushelsMessage();
                    }
                    break;

                case PLANT_SEED:

                    int acresToPlant = displayTextAndGetNumber("HOW MANY ACRES DO YOU WISH TO PLANT WITH SEED ? ");
                    if (acresToPlant < 0) {
                        gameState = GAME_STATE.FINISH_GAME;
                    }

                    if (acresToPlant <= acres) {
                        if (acresToPlant / 2 <= bushels) {
                            if (acresToPlant < 10 * population) {
                                bushels -= acresToPlant / 2;
                                peopleFed = (int) (Math.random() * 5) + 1;
                                landTradingAt = (int) peopleFed;
                                harvest = acresToPlant * landTradingAt;
                                ratsAte = 0;
                                gameState = GAME_STATE.CALCULATE_HARVEST;
                            } else {
                                notEnoughPeopleMessage();
                            }
                        } else {
                            notEnoughBushelsMessage();
                        }
                    } else {
                        notEnoughLandMessage();
                    }
                    break;

                case CALCULATE_HARVEST:

                    if ((int) (peopleFed / 2) == peopleFed / 2) {
                        // Rats are running wild
                        ratsAte = (int) (bushels / peopleFed);
                    }
                    bushels = bushels - ratsAte;
                    bushels += harvest;
                    gameState = GAME_STATE.CALCULATE_BABIES;
                    break;

                case CALCULATE_BABIES:

                    cameToCity = (int) (peopleFed * (20 * acres + bushels) / population / 100 + 1);
                    peopleFed = (bushelsToFeedPeople / 20.0d);
                    // Simplify chance of plague to a true/false
                    chanceOfPlague = (int) ((10 * (Math.random() * 2) - .3)) == 0;
                    if (population < peopleFed) {
                        gameState = GAME_STATE.YEAR_CYCLE;
                    }

                    double starved = population - peopleFed;
                    if (starved < 0.0d) {
                        starvedInAYear = 0;
                        gameState = GAME_STATE.YEAR_CYCLE;
                    } else {
                        starvedInAYear = (int) starved;
                        starvedOverall += starvedInAYear;
                        if (starved > MAX_STARVATION_IN_A_YEAR * population) {
                            starvedTooManyPeopleMessage((int) starved);
                            gameState = GAME_STATE.FINISH_GAME;
                        } else {
                            percentageStarved = ((year - 1) * percentageStarved + starved * 100 / population) / year;
                            population = (int) peopleFed;
                            gameState = GAME_STATE.YEAR_CYCLE;
                        }

                    }

                    break;


                case RESULTS:

                    int acresPerPerson = acres / population;

                    System.out.println("IN YOUR 10-YEAR TERM OF OFFICE," + String.format("%.2f", percentageStarved) + "% PERCENT OF THE");
                    System.out.println("POPULATION STARVED PER YEAR ON THE AVERAGE, I.E. A TOTAL OF");
                    System.out.println(starvedOverall + " PEOPLE DIED!!");
                    System.out.println("YOU STARTED WITH 10 ACRES PER PERSON AND ENDED WITH");
                    System.out.println(acresPerPerson + " ACRES PER PERSON.");
                    System.out.println();

                    if (percentageStarved > 33.0d || acresPerPerson < 7) {
                        starvedTooManyPeopleMessage(starvedOverall);
                    } else if (percentageStarved > 10.0d || acresPerPerson < 9) {
                        heavyHandedMessage();
                    } else if (percentageStarved > 3.0d || acresPerPerson < 10) {
                        couldHaveBeenBetterMessage();
                    } else {
                        fantasticPerformanceMessage();
                    }


                    gameState = GAME_STATE.FINISH_GAME;

                case FINISH_GAME:
                    System.out.println("SO LONG FOR NOW.");
                    gameState = GAME_STATE.GAME_OVER;

            }

        } while (gameState != GAME_STATE.GAME_OVER);
    }

    private void starvedTooManyPeopleMessage(int starved) {
        System.out.println();
        System.out.println("YOU STARVED " + starved + " PEOPLE IN ONE YEAR!!!");
        System.out.println("DUE TO THIS EXTREME MISMANAGEMENT YOU HAVE NOT ONLY");
        System.out.println("BEEN IMPEACHED AND THROWN OUT OF OFFICE BUT YOU HAVE");
        System.out.println("ALSO BEEN DECLARED NATIONAL FINK!!!!");

    }

    private void heavyHandedMessage() {
        System.out.println("DUE TO THIS EXTREME MISMANAGEMENT YOU HAVE NOT ONLY");
        System.out.println("BEEN IMPEACHED AND THROWN OUT OF OFFICE BUT YOU HAVE");
        System.out.println("ALSO BEEN DECLARED NATIONAL FINK!!!!");
    }

    private void couldHaveBeenBetterMessage() {
        System.out.println("YOUR PERFORMANCE COULD HAVE BEEN SOMEWHAT BETTER, BUT");
        System.out.println("REALLY WASN'T TOO BAD AT ALL. " + (int) (Math.random() * (population * .8)) + " PEOPLE");
        System.out.println("WOULD DEARLY LIKE TO SEE YOU ASSASSINATED BUT WE ALL HAVE OUR");
        System.out.println("TRIVIAL PROBLEMS.");
    }

    private void fantasticPerformanceMessage() {
        System.out.println("A FANTASTIC PERFORMANCE!!!  CHARLEMANGE, DISRAELI, AND");
        System.out.println("JEFFERSON COMBINED COULD NOT HAVE DONE BETTER!");
    }

    private void notEnoughPeopleMessage() {
        System.out.println("BUT YOU HAVE ONLY " + population + " PEOPLE TO TEND THE FIELDS!  NOW THEN,");

    }

    private void notEnoughBushelsMessage() {
        System.out.println("HAMURABI:  THINK AGAIN.  YOU HAVE ONLY");
        System.out.println(bushels + " BUSHELS OF GRAIN.  NOW THEN,");
    }

    private void notEnoughLandMessage() {
        System.out.println("HAMURABI:  THINK AGAIN.  YOU OWN ONLY " + acres + " ACRES.  NOW THEN,");
    }


    private void intro() {
        System.out.println(simulateTabs(32) + "HAMURABI");
        System.out.println(simulateTabs(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println();
        System.out.println("TRY YOUR HAND AT GOVERNING ANCIENT SUMERIA");
        System.out.println("FOR A TEN-YEAR TERM OF OFFICE.");
        System.out.println();
    }


    /*
     * Print a message on the screen, then accept input from Keyboard.
     * Converts input to an Integer
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private int displayTextAndGetNumber(String text) {
        return Integer.parseInt(displayTextAndGetInput(text));
    }

    /*
     * Print a message on the screen, then accept input from Keyboard.
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private String displayTextAndGetInput(String text) {
        System.out.print(text);
        return kbScanner.next();
    }

    /**
     * Simulate the old basic tab(xx) command which indented text by xx spaces.
     *
     * @param spaces number of spaces required
     * @return String with number of spaces
     */
    private String simulateTabs(int spaces) {
        char[] spacesTemp = new char[spaces];
        Arrays.fill(spacesTemp, ' ');
        return new String(spacesTemp);
    }

}

File: 43_Hammurabi/python/hamurabi.py
from random import random, seed


def gen_random():
    return int(random() * 5) + 1


def bad_input_850():
    print("\nHAMURABI:  I CANNOT DO WHAT YOU WISH.")
    print("GET YOURSELF ANOTHER STEWARD!!!!!")


def bad_input_710(S):
    print("HAMURABI:  THINK AGAIN.  YOU HAVE ONLY")
    print(S, "BUSHELS OF GRAIN.  NOW THEN,")


def bad_input_720(A):
    print("HAMURABI:  THINK AGAIN.  YOU OWN ONLY", A, "ACRES.  NOW THEN,")


def national_fink():
    print("DUE TO THIS EXTREME MISMANAGEMENT YOU HAVE NOT ONLY")
    print("BEEN IMPEACHED AND THROWN OUT OF OFFICE BUT YOU HAVE")
    print("ALSO BEEN DECLARED NATIONAL FINK!!!!")


def b_input(promptstring):  # emulate BASIC input. It rejects non-numeric values
    x = input(promptstring)
    while x.isalpha():
        x = input("?REDO FROM START\n? ")
    return int(x)


def main() -> None:
    seed()
    title = "HAMURABI"
    title = title.rjust(32, " ")
    print(title)
    attribution = "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY"
    attribution = attribution.rjust(15, " ")
    print(attribution)
    print("\n\n\n")
    print("TRY YOUR HAND AT GOVERNING ANCIENT SUMERIA")
    print("FOR A TEN-YEAR TERM OF OFFICE.\n")

    D1 = 0
    P1: float = 0
    year = 0
    population = 95
    grain_stores = 2800
    H = 3000
    eaten_rats = H - grain_stores
    bushels_per_acre = (
        3  # yield (amount of production from land). Reused as price per acre
    )
    acres = H / bushels_per_acre  # acres of land
    immigrants = 5
    plague = 1  # boolean for plague, also input for buy/sell land
    people = 0

    while year < 11:  # line 270. main loop. while the year is less than 11
        print("\n\n\nHAMURABI:  I BEG TO REPORT TO YOU")
        year = year + 1  # year
        print(
            "IN YEAR",
            year,
            ",",
            people,
            "PEOPLE STARVED,",
            immigrants,
            "CAME TO THE CITY,",
        )
        population = population + immigrants

        if plague == 0:
            population = int(population / 2)
            print("A HORRIBLE PLAGUE STRUCK!  HALF THE PEOPLE DIED.")

        print("POPULATION IS NOW", population)
        print("THE CITY NOW OWNS", acres, "ACRES.")
        print("YOU HARVESTED", bushels_per_acre, "BUSHELS PER ACRE.")
        print("THE RATS ATE", eaten_rats, "BUSHELS.")
        print("YOU NOW HAVE ", grain_stores, "BUSHELS IN STORE.\n")
        C = int(10 * random())  # random number between 1 and 10
        bushels_per_acre = C + 17
        print("LAND IS TRADING AT", bushels_per_acre, "BUSHELS PER ACRE.")

        plague = -99  # dummy value to track status
        while plague == -99:  # always run the loop once
            plague = b_input("HOW MANY ACRES DO YOU WISH TO BUY? ")
            if plague < 0:
                plague = -1  # to avoid the corner case of Q=-99
                bad_input_850()
                year = 99  # jump out of main loop and exit
            elif bushels_per_acre * plague > grain_stores:  # can't afford it
                bad_input_710(grain_stores)
                plague = -99  # give'm a second change to get it right
            elif (
                bushels_per_acre * plague <= grain_stores
            ):  # normal case, can afford it
                acres = acres + plague  # increase the number of acres by Q
                grain_stores = (
                    grain_stores - bushels_per_acre * plague
                )  # decrease the amount of grain in store to pay for it
                C = 0  # WTF is C for?

        if plague == 0 and year != 99:  # maybe you want to sell some land?
            plague = -99
            while plague == -99:
                plague = b_input("HOW MANY ACRES DO YOU WISH TO SELL? ")
                if plague < 0:
                    bad_input_850()
                    year = 99  # jump out of main loop and exit
                elif plague <= acres:  # normal case
                    acres = acres - plague  # reduce the acres
                    grain_stores = (
                        grain_stores + bushels_per_acre * plague
                    )  # add to grain stores
                    C = 0  # still don't know what C is for
                else:  # Q>A error!
                    bad_input_720(acres)
                    plague = -99  # reloop
            print("\n")

        plague = -99
        while plague == -99 and year != 99:
            plague = b_input("HOW MANY BUSHELS DO YOU WISH TO FEED YOUR PEOPLE? ")
            if plague < 0:
                bad_input_850()
                year = 99  # jump out of main loop and exit
            # REM *** TRYING TO USE MORE GRAIN THAN IS IN SILOS?
            elif plague > grain_stores:
                bad_input_710(grain_stores)
                plague = -99  # try again!
            else:  # we're good. do the transaction
                grain_stores = grain_stores - plague  # remove the grain from the stores
                C = 1  # set the speed of light to 1. jk

        print("\n")
        people = -99  # dummy value to force at least one loop
        while people == -99 and year != 99:
            people = b_input("HOW MANY ACRES DO YOU WISH TO PLANT WITH SEED? ")
            if people < 0:
                bad_input_850()
                year = 99  # jump out of main loop and exit
            elif people > 0:
                if people > acres:
                    # REM *** TRYING TO PLANT MORE ACRES THAN YOU OWN?
                    bad_input_720(acres)
                    people = -99
                elif int(people / 2) > grain_stores:
                    # REM *** ENOUGH GRAIN FOR SEED?
                    bad_input_710(grain_stores)
                    people = -99
                elif people > 10 * population:
                    # REM *** ENOUGH PEOPLE TO TEND THE CROPS?
                    print(
                        "BUT YOU HAVE ONLY",
                        population,
                        "PEOPLE TO TEND THE FIELDS!  NOW THEN,",
                    )
                    people = -99
                else:  # we're good. decrement the grain store
                    grain_stores = grain_stores - int(people / 2)

        C = gen_random()
        # REM *** A BOUNTIFUL HARVEST!
        bushels_per_acre = C
        H = people * bushels_per_acre
        eaten_rats = 0

        C = gen_random()
        if int(C / 2) == C / 2:  # even number. 50/50 chance
            # REM *** RATS ARE RUNNING WILD!!
            eaten_rats = int(
                grain_stores / C
            )  # calc losses due to rats, based on previous random number

        grain_stores = grain_stores - eaten_rats + H  # deduct losses from stores

        C = gen_random()
        # REM *** LET'S HAVE SOME BABIES
        immigrants = int(C * (20 * acres + grain_stores) / population / 100 + 1)
        # REM *** HOW MANY PEOPLE HAD FULL TUMMIES?
        C = int(plague / 20)
        # REM *** HORROS, A 15% CHANCE OF PLAGUE
        # yeah, should be HORRORS, but left it
        plague = int(10 * (2 * random() - 0.3))
        if (
            population >= C and year != 99
        ):  # if there are some people without full bellies...
            # REM *** STARVE ENOUGH FOR IMPEACHMENT?
            people = population - C
            if people > 0.45 * population:
                print("\nYOU STARVED", people, "PEOPLE IN ONE YEAR!!!")
                national_fink()
                year = 99  # exit the loop
            P1 = ((year - 1) * P1 + people * 100 / population) / year
            population = C
            D1 = D1 + people

    if year != 99:
        print("IN YOUR 10-YEAR TERM OF OFFICE,", P1, "PERCENT OF THE")
        print("POPULATION STARVED PER YEAR ON THE AVERAGE, I.E. A TOTAL OF")
        print(D1, "PEOPLE DIED!!")
        L = acres / population
        print("YOU STARTED WITH 10 ACRES PER PERSON AND ENDED WITH")
        print(L, "ACRES PER PERSON.\n")
        if P1 > 33 or L < 7:
            national_fink()
        elif P1 > 10 or L < 9:
            print("YOUR HEAVY-HANDED PERFORMANCE SMACKS OF NERO AND IVAN IV.")
            print("THE PEOPLE (REMIANING) FIND YOU AN UNPLEASANT RULER, AND,")
            print("FRANKLY, HATE YOUR GUTS!!")
        elif P1 > 3 or L < 10:
            print("YOUR PERFORMANCE COULD HAVE BEEN SOMEWHAT BETTER, BUT")
            print(
                "REALLY WASN'T TOO BAD AT ALL. ",
                int(population * 0.8 * random()),
                "PEOPLE",
            )
            print("WOULD DEARLY LIKE TO SEE YOU ASSASSINATED BUT WE ALL HAVE OUR")
            print("TRIVIAL PROBLEMS.")
        else:
            print("A FANTASTIC PERFORMANCE!!!  CHARLEMANGE, DISRAELI, AND")
            print("JEFFERSON COMBINED COULD NOT HAVE DONE BETTER!\n")
        for _ in range(1, 10):
            print("\a")

    print("\nSO LONG FOR NOW.\n")


if __name__ == "__main__":
    main()
